### PR TITLE
Adding the exact IAM message in DSS error response

### PR DESCRIPTION
**Pull request by rajatg1989**

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -1206,7 +1207,8 @@ int RGWPostObj_ObjStore_S3::get_policy()
                                                                       "",  /* Access key*/
                                                                       "",  /* Canonical string for signature */
                                                                       "", /* Received signature */
-                                                                      resource_info.getObjectName());
+                                                                      resource_info.getObjectName(),
```
Comment (shivanshu21) [inline src/rgw/rgw_rest_s3.cc:1210]: Send s->object after checking s is not null

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -1221,7 +1223,8 @@ int RGWPostObj_ObjStore_S3::get_policy()
                                                                        s3_access_key,  /* Access key */
                                                                        string(encoded_policy.c_str(),encoded_policy.length()),
                                                                        received_signature_str, /* Received signature */
-                                                                       resource_info.getObjectName());
+                                                                       resource_info.getObjectName(),
```
Comment (shivanshu21) [inline src/rgw/rgw_rest_s3.cc:1226]: s->object

File: src/rgw/rgw_rest_s3.cc [hunk 3]
```diff
@@ -2585,6 +2590,23 @@ int RGW_Auth_S3_Keystone_ValidateToken::make_iam_request(const string& keystone_
 //  std::string delimiter = "\"\"";
 //  std::string token = s.substr(3,s.find(delimiter));
   dout(0) << "DSS INFO: Printing RX headers: " << rx_headers_buffer.c_str() << dendl;
+  char *rxbuffer = strdup(rx_buffer.c_str());
```
Comment (shivanshu21) [inline src/rgw/rgw_rest_s3.cc:2593]: Why not use the safer std::string functions?

File: src/rgw/rgw_rest_s3.cc [hunk 4]
```diff
@@ -2585,6 +2590,23 @@ int RGW_Auth_S3_Keystone_ValidateToken::make_iam_request(const string& keystone_
...
+      tokens.push_back(tok);
+      p = strtok_r(NULL, "\"", &savedptr);
+  }
+
+  // assuming error from IAM is always in this format
+  // {"error": {"message": "The resource could not be found.", "code": 404, "title": "Not Found"}}
+  if(tokens.size() >= 5 && !strcmp(tokens[1].c_str(), "error") && !strcmp(tokens[3].c_str(), "message")) {
+    iamerror = "IAM_ERROR: " +  tokens[5];
+  }
+  free(rxbuffer);
```
Comment (shivanshu21) [inline src/rgw/rgw_rest_s3.cc:2609]: do we need to free savedptr?

Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc:2609]: no

**Pull request merged**